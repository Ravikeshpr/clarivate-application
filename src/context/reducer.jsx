const reducer = (state, action) => {
    switch (action.type) {
        case "ADD_DATA":
            return {
                ...state,
                list: [
                    ...new Set(
                        [...state.list, ...action.payload].map(JSON.stringify)
                    ),
                ].map(JSON.parse), //Always Unique values
            };
        case "ADD_FAVOURITE_ITEM":
            return {
                ...state,
                list: [
                    ...new Set(
                        [
                            ...state.list.toSpliced(
                                state.list.findIndex(
                                    (el) => el.id === action.payload.id
                                ),
                                1,
                                action.payload
                            ),
                        ].map(JSON.stringify)
                    ),
                ].map(JSON.parse),
            };
        case "SET_OFFSET":
            return {
                ...state,
                offset: action.payload,
            };
        case "SET_IS_API_LOAD_REQ":
            return {
                ...state,
                isAPILoadReq: action.payload,
            };
        case "SET_SCROLL_POSITION":
            return {
                ...state,
                scrollYPosition: action.payload,
            };
        default:
            return state;
    }
};

export default reducer;
