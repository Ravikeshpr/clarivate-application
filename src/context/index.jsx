/* eslint-disable react/prop-types */
import { useContext } from "react";
import { createContext, useReducer } from "react";
import reducer from "./reducer";

export const useGlobalContext = () => {
    const { state, dispatch } = useContext(globalContext);
    return { state, dispatch };
};

// Initial State
const initialState = {
    scrollYPosition: 0,
    list: [],
    offset: 1,
    isAPILoadReq: true,
};

// Create Our context
const globalContext = createContext({
    state: initialState,
    dispatch: () => {},
});

// Provider to wrap around our root react component
export const GlobalContextProvider = ({ children }) => {
    const [state, dispatch] = useReducer(reducer, initialState);
    return (
        <globalContext.Provider
            value={{
                state,
                dispatch,
            }}
        >
            {children}
        </globalContext.Provider>
    );
};
