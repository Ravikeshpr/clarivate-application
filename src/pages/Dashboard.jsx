import { useEffect } from "react";
import { Link } from "react-router-dom";
import { useGlobalContext } from "../context";
import Button from "../coponents/button/Button";

export default function Dashboard() {
    const { state, dispatch } = useGlobalContext();
    const { list } = state;
    const data = list.filter((item) => !!item.isFavourite);

    useEffect(() => {
        window.scrollTo(0, 0);
        return () => {
            if (list.length > 0) {
                dispatch({ type: "SET_IS_API_LOAD_REQ", payload: false });
            }
        };
    }, []);

    return (
        <>
            <div className="header-button">
                <Link to="/list">
                    <Button label="Go to List Page" />
                </Link>
            </div>

            <div className="container">
                <div className="row text-center mt-5">
                    {data && data.length > 0 ? (
                        <div>
                            {" "}
                            <div className="heading">
                                Favourites List Dashboard
                            </div>
                            {data.map((item) => {
                                return (
                                    <div key={item.id} className="card">
                                        <img
                                            src={item.url}
                                            alt={item.title}
                                            className="card-img"
                                        />
                                        <div className="title">
                                            <span>{item.id}:</span>
                                            <span>{` ${item.title}`}</span>
                                        </div>
                                    </div>
                                );
                            })}
                        </div>
                    ) : (
                        <div className="message">
                            No Item in the Favourites List, Go to list page and
                            add some favourites from the list
                        </div>
                    )}
                </div>
            </div>
        </>
    );
}
