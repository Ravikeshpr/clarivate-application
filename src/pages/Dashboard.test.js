import { BrowserRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import Dashboard from "./Dashboard";
import * as GlobalContex from "../context";

describe("Dashboard", () => {
    test("renders correct", () => {
        const contextValues = {
            state: {
                scrollYPosition: 0,
                list: [
                    { id: "1", title: "title", url: "url", isFavourite: true },
                ],
                offset: 1,
                isAPILoadReq: true,
            },
            dispatch: jest.fn(),
        };

        jest.spyOn(GlobalContex, "useGlobalContext").mockImplementation(
            () => contextValues
        );
        const dashboard = renderer.create(
            <BrowserRouter>
                <Dashboard />
            </BrowserRouter>
        );
        expect(dashboard).toMatchSnapshot();
    });
});
