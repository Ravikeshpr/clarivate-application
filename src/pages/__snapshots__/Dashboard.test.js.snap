// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard renders correct 1`] = `
[
  <div
    className="header-button"
  >
    <a
      href="/list"
      onClick={[Function]}
    >
      <button>
        Go to List Page
      </button>
    </a>
  </div>,
  <div
    className="container"
  >
    <div
      className="row text-center mt-5"
    >
      <div>
         
        <div
          className="heading"
        >
          Favourites List Dashboard
        </div>
        <div
          className="card"
        >
          <img
            alt="title"
            className="card-img"
            src="url"
          />
          <div
            className="title"
          >
            <span>
              1
              :
            </span>
            <span>
               title
            </span>
          </div>
          <div />
        </div>
      </div>
    </div>
  </div>,
]
`;
